core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -5874,6 +5874,11 @@ private void validateSnapshot(
         throw newValidationError(period,
             Static.RESOURCE.illegalExpressionForTemporal(dataType.getSqlTypeName().getName()));
       }
+      if (ns instanceof IdentifierNamespace && ns.resolve() instanceof WithItemNamespace) {
+        // If the snapshot is used over a CTE, then we don't have a concrete underlying
+        // table to operate on. This will be rechecked later in the planner rules.
```

Comment:
there is no guarantee that any planner rules will be executed after validation.

Reply:
i think in general you are right
I backported it from Flink where it is guaranteed...

Any proposal how to proceed with Calcite

Reply:
is SYSTEM_TIME even defined for a query?
Or just for a table?
If not, maybe the query should be rejected.

Reply:
in case of Flink it depends... It might be defined per query however then there is extra requirement of having primary key which is checked during planning.
It seems the behavior might be depending on downstream projects.

Then I'm thinking of having a `protected` method here which is failing by default in case of queries and downstream projects could override it if needed